draft-ietf-masque-h3-datagram.md

```diff
@@ -117,26 +123,192 @@ following format (using the notation from the "Notational Conventions" section
+}
+~~~
+{: #capsule-frame-format title="CAPSULE HTTP/3 Frame Format"}
+
+The Type and Length fields follows the definition of HTTP/3 frames from {{H3}}.
+The payload consists of:
+
+Capsule Type:
+
+: The type of this capsule.
+
+Capsule Data:
+
+: Data whose semantics depends on the Capsule Type.
+
+Endpoints which receive a Capsule with an unknown Capsule Type MUST silently
+drop that Capsule. Intermediaries MUST forward Capsules, even if they do not
+know the Capsule Type or cannot parse the Capsule Data.
+
+
+## The REGISTER_DATAGRAM_CONTEXT Capsule {#register-capsule}
+
+The REGISTER_DATAGRAM_CONTEXT capsule (type=0x00) allows an endpoint to inform
+its peer of the encoding and semantics of upcoming datagrams associated with
+a given context ID. Its Capsule Data field consists of:
+
+~~~
+REGISTER_DATAGRAM_CONTEXT Capsule {
+  Context ID (i),
+  Extension String (..),
+}
+~~~
+{: #register-capsule-format title="REGISTER_DATAGRAM_CONTEXT Capsule Format"}
+
+Context ID:
+
+: The context ID to register.
 
-: A variable-length integer indicating the Flow ID of the datagram (see
-{{datagram-flows}}).
+Extension String:
+
+: A string of comma-separated key-value pairs to enable extensibility.
+
+The ABNF for the Extension String field is as follows (using syntax from
+{{Section 3.2.6 of RFC7230}}):
+
+~~~
+  extension-string = [ ext-member *( "," ext-member ) ]
+  ext-member       = ext-member-key "=" ext-member-value
+  ext-member-key   = token
+  ext-member-value = token
+~~~
+
+Note that these registrations are unilateral and bidirectional: the sender of
+the frame unilaterally defines the semantics it will apply to the datagrams it
+sends and receives using this context ID. Once a context ID is registered, it
+can be used in both directions.
+
+Endpoints MUST NOT send DATAGRAM frames using a Context ID until they have
+either sent or received a REGISTER_DATAGRAM_CONTEXT Capsule with the same
+Context ID. However, due to reordering, an endpoint that receives a DATAGRAM
+frame with an unknown Context ID MUST NOT treat it as an error, it SHALL
+instead drop the DATAGRAM frame silently, or buffer it for a short while in
+hopes of receiving the corresponding REGISTER_DATAGRAM_CONTEXT Capsule.
+
+
+## The CLOSE_DATAGRAM_CONTEXT Capsule {#close-capsule}
+
+The CLOSE_DATAGRAM_CONTEXT capsule (type=0x01) allows an endpoint to inform
+its peer that it will no longer send or parse received datagrams associated with
+a given context ID. Its Capsule Data field consists of:
+
+~~~
+CLOSE_DATAGRAM_CONTEXT Capsule {
+  Context ID (i),
+}
+~~~
+{: #close-capsule-format title="CLOSE_DATAGRAM_CONTEXT Capsule Format"}
+
+Context ID:
+
+: The context ID to close.
+
+Note that this close is unilateral and bidirectional: the sender of the frame
+unilaterally informs its peer of the closure. Endpoints can use
+CLOSE_DATAGRAM_CONTEXT capsules to close a context that was initially
+registered by either themselves, or by their peer. Endpoints MAY use the
```

Comment:
Allowing endpoints to close contexts created by the peer seems confusing to me.  If my peer just closed the context I've been using ... what do I do?  Do I reopen it?  Do I fail the connection?

Reply:
I'd say that depends on the application. In CONNECT-UDP if you close the main context then things break, so you might as well close the stream. But as another example, if your peer introduced a CONNECT-IP compression context and used it for a while and you're seeing low usage of it and want to reclaim memory, you can close than context and have the tunnel still be functional. Let's move this conversation to #43.